### PR TITLE
New torch assertion macros

### PR DESCRIPTION
**Pull request by smessmer**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#20887 New torch assertion macros**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15484658/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #20821 Specify dispatch key with kernel&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D15455790/)

Switch AT_xxx assertion macros to the TORCH_ variants and make sure the separation between TORCH_CHECK and TORCH_INTERNAL_ASSERT makes sense.

Differential Revision: [D15484658](https://our.internmc.facebook.com/intern/diff/D15484658/)

### CODE REVIEW

**Comment by ezyang** [comment]
I only reviewed a few of the sites, but here's a hot take: nearly all of these should be `TORCH_CHECK`. Let's treat the dispatcher table as an abstraction barrier, so you can use asserts for internal invariants that shouldn't be breakable by the API. But there is definitely a public API provided by the dispatcher table, and most of these things are checks for misuses of that API.)

(The other test: if I deleted all of your `TORCH_INTERNAL_ASSERT` lines, would you still be happy?)

**Review by ezyang — CHANGES_REQUESTED** [review]
getting it out of my queue

**Comment by smessmer** [comment]
@ezyang The dispatch table class itself isn't meant to be part of the public API. The registration API is. All cases that could fail by using the registration API wrongly are `TORCH_CHECK`. The `TORCH_INTERNAL_ASSERT`s check things that can't be user errors like if the registration API destructor tries to delete something it didn't register before.

File: aten/src/ATen/core/op_registration/op_registration.cpp [hunk 1]
```diff
@@ -15,7 +15,7 @@ class RegisterOperators::OperatorRegistrar final {
   explicit OperatorRegistrar(FunctionSchema&& schema, c10::optional<TensorTypeId> dispatch_key, KernelFunction* kernel, KernelCacheCreatorFunction&& cache_creator)
   : op_(Dispatcher::singleton().registerSchema(std::move(schema))), kernel_registration_handle_(c10::nullopt) {
     // either both, kernel and cache_creator, or none must be set.
-    AT_ASSERT((kernel != nullptr) == static_cast<bool>(cache_creator));
+    TORCH_INTERNAL_ASSERT((kernel != nullptr) == static_cast<bool>(cache_creator));
```
Comment (ezyang) [inline aten/src/ATen/core/op_registration/op_registration.cpp]: This is in c10 namespace, so I guess it is public API. Isn't this assert invalid?

File: aten/src/ATen/core/op_registration/op_registration.cpp [hunk 2]
```diff
@@ -134,7 +134,7 @@ void RegisterOperators::registerOp_(FunctionSchema&& schema, Options&& options)
 }
 
 void RegisterOperators::registerSchemaAndKernel_(FunctionSchema schema, Options::KernelRegistrationConfig&& kernel) {
-  AT_ASSERTM(kernel.kernel_func != nullptr && static_cast<bool>(kernel.cache_creator_func), "Kernel must be set");
+  TORCH_INTERNAL_ASSERT(kernel.kernel_func != nullptr && static_cast<bool>(kernel.cache_creator_func), "Kernel must be set");
```
Comment (ezyang) [inline aten/src/ATen/core/op_registration/op_registration.cpp]: Similarly here

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/core/op_registration/op_registration.cpp (hunk 1 above)
Comment (smessmer) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: This class is defined in a .cpp file, not in a header. Anything outside this .cpp file can't call it.

File: aten/src/ATen/core/op_registration/op_registration.cpp (hunk 2 above)
Comment (smessmer) [reply aten/src/ATen/core/op_registration/op_registration.cpp]: `registerSchemaAndKernel_` is a private method. If, by using public methods, you're triggering this assert, then that is a pytorch bug.

**Pull request closed without merging**